File: ClanTabKit/Sources/ClanTabKit/Logic/Validation.swift
import Foundation

/// Errors surfaced by `Validation`. The server-side DO validates independently of the
/// UI — see `DESIGN.md` §6 — so these are the same checks a request handler would run.
public enum ValidationError: Error, Equatable, Sendable {
    case emptySplits
    case splitMismatch(expected: Int64, actual: Int64)
    case unknownMember(String)
    case invalidAmount(Int64)
    /// An itemized expense with no line items.
    case emptyItems
    /// A line item with no-one sharing it.
    case itemWithoutParticipants
    /// The line items don't sum to the expense amount.
    case itemSumMismatch(expected: Int64, actual: Int64)
    /// A `shares` expense with no weights.
    case emptyShares
    /// A share weight is negative, or every weight is `0` (nothing to divide by).
    case invalidShareWeight
    /// An expense with no payers — every expense needs at least one.
    case emptyPayers
    /// The payers' contributions don't sum to the expense amount.
    case payerSumMismatch(expected: Int64, actual: Int64)
}

/// Split-sum validation and deterministic remainder distribution.
public enum Validation {
    /// Verifies that `splits` sum exactly to `amountMinor`. No tolerance: an equal
    /// split with a remainder must already have that remainder deterministically
    /// assigned (see `equalSplit`) before this is called.
    public static func validateSplitsSum(amountMinor: Int64, splits: [ExpenseSplit]) throws {
        guard !splits.isEmpty else {
            throw ValidationError.emptySplits
        }
        let total = splits.reduce(Int64(0)) { $0 + $1.amountMinor }
        guard total == amountMinor else {
            throw ValidationError.splitMismatch(expected: amountMinor, actual: total)
        }
    }

    /// Verifies that `payers` sum exactly to `amountMinor` (`CHECKLIST.md`
    /// "Multiple payers on one expense") — the same integrity rule
    /// `validateSplitsSum` already enforces, mirrored for the credit side of
    /// an expense rather than the debit side.
    public static func validatePayersSum(amountMinor: Int64, payers: [ExpensePayment]) throws {
        guard !payers.isEmpty else {
            throw ValidationError.emptyPayers
        }
        let total = payers.reduce(Int64(0)) { $0 + $1.amountMinor }
        guard total == amountMinor else {
            throw ValidationError.payerSumMismatch(expected: amountMinor, actual: total)
        }
    }

    /// Verifies every referenced member id (payer, split members, settlement
    /// from/to) actually belongs to the group.
    public static func validateMembersExist(memberIds: [String], validMemberIds: Set<String>) throws {
        for id in memberIds {
            guard validMemberIds.contains(id) else {
                throw ValidationError.unknownMember(id)
            }
        }
    }

    /// Amounts are always positive integers in minor units.
    public static func validatePositiveAmount(_ amountMinor: Int64) throws {
        guard amountMinor > 0 else {
            throw ValidationError.invalidAmount(amountMinor)
        }
    }

    /// Divides `amountMinor` evenly across `memberIds`, assigning the integer-division
    /// remainder to `remainderRecipient` (normally the payer) so the result always sums
    /// to exactly `amountMinor` — never gaining or losing a paisa. Falls back to the
    /// first member if `remainderRecipient` isn't among `memberIds`, so the sum
    /// invariant holds unconditionally.
    public static func equalSplit(
        amountMinor: Int64,
        memberIds: [String],
        remainderRecipient: String
    ) -> [ExpenseSplit] {
        precondition(!memberIds.isEmpty, "Cannot split an expense among zero members")

        let count = Int64(memberIds.count)
        let base = amountMinor / count
        let remainder = amountMinor % count
        let recipient = memberIds.contains(remainderRecipient) ? remainderRecipient : memberIds[0]

        return memberIds.map { id in
            ExpenseSplit(memberId: id, amountMinor: id == recipient ? base + remainder : base)
        }
    }

    /// Divides `amountMinor` in proportion to integer `weights` (percentages or
    /// shares — any non-negative integers; they need not sum to 100). Each member
    /// gets `floor(amountMinor * weight / totalWeight)`; the leftover minor units
    /// — always fewer than the member count — are assigned to `remainderRecipient`
    /// (normally the payer), falling back to the first member, so the result sums
    /// to exactly `amountMinor`, never gaining or losing a paisa. A `0` weight
    /// yields a `0` share (the member is on the expense but owes nothing for it).
    ///
    /// This is the `percentage` counterpart to `equalSplit`: the UI works in
    /// percentages, the wire only ever carries resolved minor-unit splits.
    public static func percentageSplit(
        amountMinor: Int64,
        weights: [(memberId: String, weight: Int)],
        remainderRecipient: String
    ) -> [ExpenseSplit] {
        precondition(!weights.isEmpty, "Cannot split an expense among zero members")
        let totalWeight = weights.reduce(Int64(0)) { $0 + Int64(max(0, $1.weight)) }
        precondition(totalWeight > 0, "Percentage split needs at least one positive weight")

        var splits = weights.map { entry in
            ExpenseSplit(
                memberId: entry.memberId,
                amountMinor: amountMinor * Int64(max(0, entry.weight)) / totalWeight
            )
        }

        let allocated = splits.reduce(Int64(0)) { $0 + $1.amountMinor }
        let remainder = amountMinor - allocated
        if remainder != 0 {
            let index = splits.firstIndex { $0.memberId == remainderRecipient } ?? 0
            splits[index] = ExpenseSplit(
                memberId: splits[index].memberId,
                amountMinor: splits[index].amountMinor + remainder
            )
        }
        return splits
    }

    /// The `shares` counterpart to `percentageSplit` — identical maths (both
    /// divide `amountMinor` by arbitrary non-negative integer weights, leftover
    /// minor units to `remainderRecipient`), named separately only so call sites
    /// read clearly. The weights are raw ratios (4 : 2 : 1), not percentages.
    public static func sharesSplit(
        amountMinor: Int64,
        weights: [(memberId: String, weight: Int)],
        remainderRecipient: String
    ) -> [ExpenseSplit] {
        percentageSplit(amountMinor: amountMinor, weights: weights, remainderRecipient: remainderRecipient)
    }

    /// Checks a `shares` expense's weights: at least one weight, none negative,
    /// a positive total (something to divide by), and every member a real group
    /// member. The weights themselves are unconstrained ratios.
    public static func validateShares(
        weights: [ShareWeight],
        validMemberIds: Set<String>
    ) throws {
        guard !weights.isEmpty else { throw ValidationError.emptyShares }
        guard weights.allSatisfy({ $0.weight >= 0 }) else { throw ValidationError.invalidShareWeight }
        guard weights.reduce(0, { $0 + $1.weight }) > 0 else { throw ValidationError.invalidShareWeight }
        try validateMembersExist(memberIds: weights.map(\.memberId), validMemberIds: validMemberIds)
    }

    /// Resolves an itemized expense (`LineItem`s) into one `ExpenseSplit` per
    /// member who's on at least one item. Each item is divided equally among its
    /// own participants — `equalSplit`'s exact rule, with that item's own
    /// leftover minor unit(s) going to `remainderRecipient` (the payer) when
    /// they're a participant, else the item's first participant — and each
    /// member's shares are then summed across every item.
    ///
    /// If the items sum to `amountMinor` (which `validateItems` enforces before
    /// this is relied on), the returned splits sum to `amountMinor` exactly,
    /// because every item resolves exactly. Members not on any item are omitted
    /// entirely rather than carried at `0` — they simply aren't part of the
    /// expense.
    ///
    /// This is the `itemized` counterpart to `equalSplit` / `percentageSplit`:
    /// the UI works in line items, the wire only ever carries resolved
    /// minor-unit splits (`DESIGN.md` §6).
    public static func itemizedSplit(
        items: [LineItem],
        remainderRecipient: String
    ) -> [ExpenseSplit] {
        var totals: [String: Int64] = [:]
        var order: [String] = []

        for item in items where !item.participantIds.isEmpty {
            let itemSplits = equalSplit(
                amountMinor: item.amountMinor,
                memberIds: item.participantIds,
                remainderRecipient: remainderRecipient
            )
            for split in itemSplits {
                if totals[split.memberId] == nil { order.append(split.memberId) }
                totals[split.memberId, default: 0] += split.amountMinor
            }
        }

        return order.map { ExpenseSplit(memberId: $0, amountMinor: totals[$0] ?? 0) }
    }

    /// Checks an itemized expense's line items: at least one item, every item
    /// shared by at least one member, every participant a real group member, and
    /// the item amounts summing to exactly `amountMinor` (no separate tax/tip
    /// bucket — a shared surcharge is its own line). Individual item amounts must
    /// be positive.
    public static func validateItems(
        amountMinor: Int64,
        items: [LineItem],
        validMemberIds: Set<String>
    ) throws {
        guard !items.isEmpty else { throw ValidationError.emptyItems }

        var total: Int64 = 0
        for item in items {
            guard item.amountMinor > 0 else { throw ValidationError.invalidAmount(item.amountMinor) }
            guard !item.participantIds.isEmpty else { throw ValidationError.itemWithoutParticipants }
            try validateMembersExist(memberIds: item.participantIds, validMemberIds: validMemberIds)
            total += item.amountMinor
        }
        guard total == amountMinor else {
            throw ValidationError.itemSumMismatch(expected: amountMinor, actual: total)
        }
    }
}

File: worker/src/lib/validation.ts
import type { ExpensePayment, ExpenseSplit, LineItem, ShareWeight } from "./types.ts";

// Money is integer minor units. JavaScript `number` is exact for integers up to
// 2^53; a group's running totals stay many orders of magnitude below that, so
// `number` is safe here as long as every amount is guarded as a positive integer
// on the way in. Reject anything non-integer rather than truncating.

export type ValidationCode =
  | "SPLIT_MISMATCH"
  | "UNKNOWN_MEMBER"
  | "INVALID_AMOUNT"
  // edit / delete: the expense or settlement id in the path doesn't exist
  | "NOT_FOUND"
  // a member can't be removed — it has activity, is claimed, or is the last one
  | "MEMBER_IN_USE"
  // accounts / claim flow (ACCOUNTS_DESIGN.md §6)
  | "ALREADY_CLAIMED"
  | "IDENTITY_ALREADY_IN_GROUP";

/** A validation failure that maps directly to a `DESIGN.md` §2 error envelope. */
export class ValidationFailure extends Error {
  constructor(
    public readonly code: ValidationCode,
    message: string,
  ) {
    super(message);
    this.name = "ValidationFailure";
  }
}

/** Amounts are always positive integers in minor units. */
export function assertPositiveAmount(amountMinor: unknown): asserts amountMinor is number {
  if (typeof amountMinor !== "number" || !Number.isInteger(amountMinor) || amountMinor <= 0) {
    throw new ValidationFailure("INVALID_AMOUNT", "Amount must be a positive integer in minor units.");
  }
}

/**
 * Splits must sum to exactly `amountMinor` — no tolerance. An `equal` split with a
 * remainder is expected to have had that remainder deterministically assigned
 * (client-side, to the payer) before the request is sent, so this is always an
 * exact check (`DESIGN.md` §6). Matches `ClanTabKit`'s `Validation.validateSplitsSum`:
 * only non-empty + exact sum are checked. Individual shares may be `0` (a member
 * included in the expense who owes nothing for it, e.g. `1` minor unit split three
 * ways → `1, 0, 0`); they must be non-negative integers but need not be positive.
 */
export function assertSplitsSum(amountMinor: number, splits: ExpenseSplit[]): void {
  if (splits.length === 0) {
    throw new ValidationFailure("SPLIT_MISMATCH", "An expense must have at least one split.");
  }
  for (const s of splits) {
    if (typeof s.amountMinor !== "number" || !Number.isInteger(s.amountMinor) || s.amountMinor < 0) {
      throw new ValidationFailure("SPLIT_MISMATCH", "Each split amount must be a non-negative integer.");
    }
  }
  const total = splits.reduce((acc, s) => acc + s.amountMinor, 0);
  if (total !== amountMinor) {
    throw new ValidationFailure(
      "SPLIT_MISMATCH",
      `Splits sum to ${total} but the expense amount is ${amountMinor}.`,
    );
  }
}

/**
 * Payers must sum to exactly `amountMinor` — the same integrity rule
 * `assertSplitsSum` already enforces, mirrored for the credit side of an
 * expense (`CHECKLIST.md` "Multiple payers on one expense"). Matches
 * `ClanTabKit`'s `Validation.validatePayersSum`.
 */
export function assertPayersSum(amountMinor: number, payers: ExpensePayment[]): void {
  if (payers.length === 0) {
    throw new ValidationFailure("SPLIT_MISMATCH", "An expense must have at least one payer.");
  }
  for (const p of payers) {
    if (typeof p.amountMinor !== "number" || !Number.isInteger(p.amountMinor) || p.amountMinor <= 0) {
      throw new ValidationFailure("INVALID_AMOUNT", "Each payer's contribution must be a positive integer.");
    }
  }
  const total = payers.reduce((acc, p) => acc + p.amountMinor, 0);
  if (total !== amountMinor) {
    throw new ValidationFailure(
      "SPLIT_MISMATCH",
      `Payers sum to ${total} but the expense amount is ${amountMinor}.`,
    );
  }
}

/** Every referenced member id (payer, split members, settlement from/to) must exist. */
export function assertMembersExist(referencedIds: readonly string[], groupMemberIds: ReadonlySet<string>): void {
  for (const id of referencedIds) {
    if (!groupMemberIds.has(id)) {
      throw new ValidationFailure("UNKNOWN_MEMBER", `Member "${id}" is not in this group.`);
    }
  }
}

/**
 * The line items of an `itemized` expense (`FEATURE_BACKLOG.md` "Itemized
 * expense entry"). At least one item; every item a positive integer amount,
 * shared by at least one real group member; and the item amounts summing to
 * exactly `amountMinor` — there is no separate tax/tip bucket, a shared
 * surcharge is entered as its own line. Mirrors `ClanTabKit.Validation.validateItems`.
 * The resolved `splits` are checked separately by `assertSplitsSum`.
 */
export function assertItemsValid(
  amountMinor: number,
  items: LineItem[],
  groupMemberIds: ReadonlySet<string>,
): void {
  if (items.length === 0) {
    throw new ValidationFailure("SPLIT_MISMATCH", "An itemized expense must have at least one line item.");
  }
  let total = 0;
  for (const item of items) {
    if (typeof item.amountMinor !== "number" || !Number.isInteger(item.amountMinor) || item.amountMinor <= 0) {
      throw new ValidationFailure("INVALID_AMOUNT", "Each line item amount must be a positive integer in minor units.");
    }
    if (item.participantIds.length === 0) {
      throw new ValidationFailure("SPLIT_MISMATCH", `Line item "${item.name}" has no-one sharing it.`);
    }
    assertMembersExist(item.participantIds, groupMemberIds);
    total += item.amountMinor;
  }
  if (total !== amountMinor) {
    throw new ValidationFailure(
      "SPLIT_MISMATCH",
      `Line items sum to ${total} but the expense amount is ${amountMinor}.`,
    );
  }
}

/**
 * The ratio weights of a `shares` expense (`CHECKLIST.md` "Split by shares").
 * At least one weight; every weight a non-negative integer, for a real group
 * member; and a positive total (something to divide by). The weights themselves
 * are arbitrary ratios and are not otherwise constrained. Mirrors
 * `ClanTabKit.Validation.validateShares`. The resolved `splits` are checked
 * separately by `assertSplitsSum`.
 */
export function assertSharesValid(
  shares: ShareWeight[],
  groupMemberIds: ReadonlySet<string>,
): void {
  if (shares.length === 0) {
    throw new ValidationFailure("SPLIT_MISMATCH", "A shares expense must have at least one weight.");
  }
  let total = 0;
  for (const share of shares) {
    if (typeof share.weight !== "number" || !Number.isInteger(share.weight) || share.weight < 0) {
      throw new ValidationFailure("INVALID_AMOUNT", "Each share weight must be a non-negative integer.");
    }
    total += share.weight;
  }
  if (total === 0) {
    throw new ValidationFailure("SPLIT_MISMATCH", "Share weights add up to zero — nothing to divide by.");
  }
  assertMembersExist(shares.map((s) => s.memberId), groupMemberIds);
}
